### ISSUE TITLE
Import of secp384r1 private key in isoApplet card fails

### ISSUE THREAD

**Issue by jo-bitsch** [OpenSC/OpenSC#1747]
### Problem Description

When trying to import a secp384r1 private key into an isoApplet card using pkcs15-init:
`pkcs15-init -S test.key  -i 1 -u sign -l tls -a ff`
I get an error message:
`Failed to store private key: Invalid arguments`

Digging deeper into this, it turns out that the apdu len was 357 bytes but isoApplet_put_data_prkey_ec tried to send it as a short apdu, enven though the card didn't allow extended apdus.

### Proposed Resolution

Enable APDU chaining if the length of the apdu is greater than 255 bytes and the card does not allow extended APDUs in isoApplet_put_data_prkey_ec, as is already checked in isoApplet_put_data_prkey_rsa.

I am preparing a pull request for this.

### Steps to reproduce

Create a EC private key on the NIST P-384 curve and try to import it onto a card with an isoApplet which does not have extended APDUs enabled.

```
openssl genpkey -algorithm EC -pkeyopt ec_paramgen_curve:P-384 -pkeyopt ec_param_enc:named_curve -out test.key
pkcs15-init -S test.key  -a ff
```

### Logs
```
0x7f2d088ad740 16:20:38.256 [pkcs15-init] card-isoApplet.c:872:isoApplet_put_data_prkey_ec: called
0x7f2d088ad740 16:20:38.256 [pkcs15-init] card-isoApplet.c:505:isoApplet_put_ec_params: called
0x7f2d088ad740 16:20:38.256 [pkcs15-init] card-isoApplet.c:538:isoApplet_put_ec_params: returning with: 0 (Success)
0x7f2d088ad740 16:20:38.256 [pkcs15-init] apdu.c:554:sc_transmit_apdu: called
0x7f2d088ad740 16:20:38.256 [pkcs15-init] apdu.c:261:sc_check_apdu: failed length check for short APDU
0x7f2d088ad740 16:20:38.256 [pkcs15-init] apdu.c:339:sc_check_apdu: Invalid Case 3 short APDU:
cse=03 cla=00 ins=db p1=3f p2=ff lc=357 le=0
resp=(nil) resplen=0 data=0x7ffca021b990 datalen=357
0x7f2d088ad740 16:20:38.256 [pkcs15-init] card-isoApplet.c:934:isoApplet_put_data_prkey_ec: APDU transmit failed
0x7f2d088ad740 16:20:38.256 [pkcs15-init] card-isoApplet.c:967:isoApplet_put_data_prkey_ec: returning with: -1300 (Invalid arguments)
```

**Comment by mouse07410** [comment]
There's something I'm missing here: how can a private key that's 384 **bits** or less, not fit into 200+ **bytes**?  

A private EC key over secp384r1 is just an integer of at most 48 bytes.

**Comment by jo-bitsch** [comment]
You are assuming that only the private key and the curve name are being transferred. However, this is not the case. The curve parameters are transferred explicitly with the private key which includes (including the TLV tags):
 *  0xE0 - Private class, constructed encoding, number one.
 *  0x81 - prime
 *  0x82 - coefficient A
 *  0x83 - coefficient B
 *  0x84 - base point G
 *  0x85 - order
 *  0x87 - cofactor
 *  0x88 - private D (private key)

(7 of these are in the same magnitude as the private key itself in length)

**Comment by dengert** [comment]
https://github.com/OpenSC/OpenSC/pull/1734 may fix the problem. The  code being patched assumed  extended. The patch  will use data chaining with short APDUs.

**Comment by dengert** [comment]
Actually #1734 will not fix the problem, as card-isoApplet.c is not using sc_format_apdu_cse_lc_le yet. 

But  has code to handle using data chaining in  `isoApplet_put_data_prkey_rsa` lines 811-815, but needs the same code in `isoApplet_put_data_prkey_ec`before line 930. Something like this untested fix:

[fix-1747.diff.txt](https://github.com/OpenSC/OpenSC/files/3414919/fix-1747.diff.txt)

**Comment by llogar** [comment]
Also note that support for `secp384r1` is missing in `card-isoApplet.c` and `pkcs15-isoApplet.c`, so you won't be able to do much with the key... I fixed this (see commit https://github.com/llogar/OpenSC/commit/5255b2143f44087d505660af8ccec42c6a01f3b5, and, @dengert, yes `fix-1747.diff` is correct) and lots of other changes/improvements to both IsoApplet (see https://github.com/llogar/IsoApplet/tree/v3) and related OpenSC (see https://github.com/llogar/OpenSC/commits/v5) files. I've also notified Philip Wendland (author of IsoApplet) of my work, but it hasn't been merged (yet?)...

**Comment by jo-bitsch** [comment]
I think `fix-1747.diff` should also include a length check on the apdu, like this:
```
	if ((apdu.datalen > 255) && !(card->caps & SC_CARD_CAP_APDU_EXT))
	{
		apdu.flags |= SC_APDU_FLAGS_CHAINING;
	}
```

I tested this on my linux machine and a javacard with the isoApple and the key import worked.

Using the pkcs15-crypt tool I'm able to use the key
```
pkcs15-crypt -i textToSign.txt -s
```
also with
```
pkcs11-tool --id 1 -s -m "ECDSA-SHA1" --input-file textToSign.txt --output-file sigtest
```

I prepared a pull reqest: #1748 

Using the card in openssl is a different story, as the isoApplet currently only supports `ALG_ECDSA_SHA1`. I didn't have the chance to try out @llogar's changes yet, which seem to fix it.

**Comment by llogar** [comment]
@jo-bitsch I believe my patches are also needed if you want to use on-card key generation...

**Comment by philipWendland** [comment]
> 
> 
> Also note that support for `secp384r1` is missing in `card-isoApplet.c` and `pkcs15-isoApplet.c`, so you won't be able to do much with the key... I fixed this (see commit [llogar@5255b21](https://github.com/llogar/OpenSC/commit/5255b2143f44087d505660af8ccec42c6a01f3b5), and, @dengert, yes `fix-1747.diff` is correct) and lots of other changes/improvements to both IsoApplet (see https://github.com/llogar/IsoApplet/tree/v3) and related OpenSC (see https://github.com/llogar/OpenSC/commits/v5) files. I've also notified Philip Wendland (author of IsoApplet) of my work, but it hasn't been merged (yet?)...

Hi,
the secp384r1 was already in the card-isoApplet.c code. Your unmerged changes added it (incorrectly) a second time. So the fix should work without your patch.

@llogar, your additional features have not been merged yet. It is a large change - currently I need to allocate most of my spare time for another project. But it is on my list.
